source/controllers/include/controllers/Controller.hpp

```diff
@@ -0,0 +1,72 @@
+/**
+ * @author Baptiste Busch
+ * @date 2021/02
+ *
+ */
+
+#pragma once
+
+#include "controllers/exceptions/NotImplementedException.hpp"
+#include "state_representation/Parameters/ParameterInterface.hpp"
+#include <list>
+#include <memory>
+
+namespace controllers {
+/**
+ * @class Controller
+ * @brief Abstract class to define a controller either in joint or cartesian spaces
+ * @tparam SIn the input state type of the controller
+ * @tparam SOut the output command type of the controller
+ */
+template <class SIn, class SOut>
+class Controller {
+public:
+  /**
+   * @brief Empty constructor
+   */
+  explicit Controller();
+
+  /**
+   * @brief Compute the command based on the input state
+   * To be redefined based on the actual controller implementation
+   * @param state the input state of the system. This function accept any number of extra arguments.
+   * @return the output command at the input state
+   */
+  virtual SOut compute_command(const SIn& state, ...) const;
+
+  /**
+   * @brief Compute the command based on the desired state and a feedback state
+   * To be redefined based on the actual controller implementation.
+   * This function accept any number of extra arguments.
+   * @param desired_state the desired state of the system.
+   * @param feedback_state the real state of the system as read from feedback loop
+   * @return the output command at the input state
+   */
+  virtual SOut compute_command(const SIn& desired_state, const SIn& feedback_state) const;
+
+  /**
+   * @brief Return a list of all the parameters of the controller
+   * @return the list of parameters
+   */
+  virtual std::list<std::shared_ptr<StateRepresentation::ParameterInterface>> get_parameters() const;
+};
+
+template <class SIn, class SOut>
+Controller<SIn, SOut>::Controller() {}
+
+template <class SIn, class SOut>
+std::list<std::shared_ptr<StateRepresentation::ParameterInterface>> Controller<SIn, SOut>::get_parameters() const {
+  std::list<std::shared_ptr<StateRepresentation::ParameterInterface>> param_list;
+  return param_list;
+}
+
+template <class SIn, class SOut>
+SOut Controller<SIn, SOut>::compute_command(const SIn&, ...) const {
+  throw exceptions::NotImplementedException("compute_command(const SIn& state, ...) not implemented for the base controller class");
```

Comment:
Just wondering, why do you prefer this pattern with "not implemented" over using pure virtual functions? I guess so that you can actually compile code / test code with the base class or other still abstract derived classes?

Reply:
Mainly because I am not sure that a derived class not to redefine all the possibilities. If those are virtual pure then it become compulsory and if they are not defined in the base class there is an error at linkage